src/main/java/com/eerussianguy/blazemap/gui/BlazeGui.java

```diff
@@ -38,73 +39,80 @@ protected BlazeGui(@Nonnull Component title, int guiWidth, int guiHeight) {
         this.top = (height - guiHeight) / 2;
+
+        addRenderableOnly(this.widgetFrame);
     }
 
     @Override
     public void render(GuiGraphics graphics, int i0, int i1, float f0) {
         PoseStack stack = graphics.pose();
-        
-        renderBackground(graphics);
 
-        var buffers = MultiBufferSource.immediate(Tesselator.getInstance().getBuilder());
-        renderFrame(graphics, buffers);
-
-        stack.pushPose();
-        stack.translate(left, top, 0.05F);
-        if(title != EMPTY) {
-            renderLabel(graphics, buffers, title, 12, 12, true);
-        }
-        renderComponents(graphics, buffers);
-        stack.popPose();
-        buffers.endBatch();
+        // Render the dark overlay over the game
+        renderBackground(graphics);
 
+        // This will render all the Renderable widgets declared in the init function
         stack.pushPose();
         stack.translate(0, 0, 0.1F);
         super.render(graphics, i0, i1, f0);
         stack.popPose();
 
+        // For things outside the GUI frame such as the editable minimap
         stack.pushPose();
         float scale = (float) getMinecraft().getWindow().getGuiScale();
         float unscale = 1F / scale;
+
         stack.scale(unscale, unscale, 1);
         stack.translate(0, 0, 0.5F);
-        buffers = MultiBufferSource.immediate(Tesselator.getInstance().getBuilder());
-        renderAbsolute(graphics, buffers, scale);
-        buffers.endBatch();
-        stack.popPose();
-    }
-
-    protected void renderAbsolute(GuiGraphics graphics, MultiBufferSource buffers, float scale){}
 
-    protected void renderFrame(GuiGraphics graphics, MultiBufferSource buffers) {
-        PoseStack stack = graphics.pose();
-
-        stack.pushPose();
-        stack.translate(left, top, 0);
-        RenderHelper.drawFrame(buffers.getBuffer(background), graphics, guiWidth, guiHeight, 8);
+        renderAbsolute(graphics, scale);
         stack.popPose();
     }
 
-    protected abstract void renderComponents(GuiGraphics graphics, MultiBufferSource buffers);
+    protected void renderAbsolute(GuiGraphics graphics, float scale){}
 
-    protected void renderLabel(GuiGraphics graphics, MultiBufferSource buffers, Component text, int x, int y, boolean shadow) {
+    protected void renderLabel(GuiGraphics graphics, Component text, int x, int y, boolean shadow) {
         PoseStack stack = graphics.pose();
+        MultiBufferSource buffers = graphics.bufferSource();
         this.font.drawInBatch(text, x, y, shadow ? Colors.WHITE : Colors.LABEL_COLOR, shadow, stack.last().pose(), buffers, Font.DisplayMode.NORMAL, 0, LightTexture.FULL_BRIGHT);
     }
 
-    protected void renderSlot(GuiGraphics graphics, MultiBufferSource buffers, int x, int y, int width, int height) {
-        PoseStack stack = graphics.pose();
+    /**
+     * This is the shared frame that forms the background of all GUI widgets
+     */
+    protected class BlazeWidgetFrame implements Renderable {
```

Comment:
Why does this need to be its own class?

Reply:
A widget needs to extend `Renderable`. Needed to do this to be able to add the frame as a widget, which was required to get it to render on the Screen over the background layer. 

Should probably exend the `GuiPrimitive` class to have all the functionality of the other elements I added, but eh I made this before I made that, and making the others all nest inside this was a bigger refactor than I wanted to tackle at this stage (was trying to _not_ do too much of BME-82 given the chance of you wanting to do it completely differently 😛).